[Step2] 연관 관계 매핑

7271kim opened the pull request:
안녕하세요! ssosso 리뷰어님~
2단계 - 연관 관계 매핑 작업 완료하여 피드백 요청드립니다.
우선 힌트에 제시된 DDL 기반으로 작업하였고, answer와 question은 answer가 주인이면서 양방향 관계가 되도록 작업하였습니다!
보시면서 개선해야 할 점이나, 추가하면 좋을 부분 있다면 코멘트 부탁드립니다~

7271kim committed:
[Step2] 연관 관계 매핑

Answer 연관관계 수정
alter table answer
    add constraint fk_answer_to_question
        foreign key (question_id)
            references question
alter table answer
    add constraint fk_answer_writer
        foreign key (writer_id)
            references user

7271kim committed:
[Step2] 연관 관계 매핑 - DeleteHistory

DeleteHistory 연관관계 수정

alter table delete_history
    add constraint fk_delete_history_to_user
        foreign key (deleted_by_id)
            references user

7271kim committed:
[Step2] 연관 관계 매핑 - Question

Question 연관관계 수정
alter table question
    add constraint fk_question_writer
        foreign key (writer_id)
            references user

7271kim committed:
[Step2] 연관 관계 매핑 - 관련  Service 수정

getWriterId > getWriter 변경

7271kim committed:
fix : Answer, Question 양방향 관계 수정

방어로직 및 양방향 세팅

ssosso commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -1,66 +1,66 @@
+
+        List<DeleteHistory> deleteHistories = new ArrayList<>();
+        question.setDeleted(true);
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, question.getWriter()));
```
Question, Answer에게 DeleteHistory를 만들어 반환하라고 하면 어떨까요~? getter와 setter를 없앨 수 있을까요~?

ssosso reviewed (changes requested):
석진님~ 피드백 드리오니 확인 부탁드립니다:)

7271kim committed:
refator : [Step2] deleteQuestion() 메서드 단순화

- Question, Answer에게 DeleteHistory 생성 책임 전가.
- 불필요한 Setter 제거.
- Java Exception Handle in Stream Operations 적용 ( 참고 주소 )
https://www.oreilly.com/content/handling-checked-exceptions-in-java-streams/

7271kim commented:
안녕하세요 소현님!
기존에 존재하였던 deleteQuestion() 메서드 관련 리펙토링 진행 완료하였습니다 :)
말씀 하신 것 처럼 Question, Answer에서 각각  삭제된 DeleteHistory 를 생성 하여 return 하도록 대한 책임을 전가하였습니다.
말씀 해주신 대로 작업하다 보니 훨씬 코드가 깔끔해 진 것 같아 기분이 좋네요. 책임 전가로 인한 불필요한 getter, setter 제거 또한 진행했습니다.
확인 부탁드립니다.
감사합니다!

ssosso commented on src/main/java/qna/domain/Question.java:
```diff
@@ -97,16 +95,42 @@ public boolean isDeleted() {
+        return answers.stream().map(wrap(answer -> answer.deleteByOwner(loginUser))).collect(Collectors.toList());
+    }
+
+    public static <T, R, E extends Exception> Function<T, R> wrap(ExceptionFunction<T, R, E> function) {
```
혹시 wrap을 사용하게 된 이유가 있을까요~?
wrap이라는 메서드를 내부에서 사용하기는 하지만 Question에 public static으로 외부에 기능을 제공하는 것이 적절한 것인지 확인 부탁드립니다:)
그리고 CannotDeleteException이 지금 보니 Exception을 상속받았던데 exception 발생 시 @transactional rollback이 제대로 동작할 수 있을까요~?

ssosso reviewed (changes requested):
석진님:)~ 피드백 드리오니 확인 부탁드립니다~

7271kim committed:
fix : CannotDeleteException 상속 수정

Exception > Runtime Exception으로 수정

7271kim commented:
안녕하세요. 소현님!
질문에 우선 답변드리겠습니다 :)
wrap을 사용하게 된 이유는 CannotDeleteException.java 파일이 초기 fork시 부터 Exception을 상속받고 있어 stream 안에 try catch문이 생기게 되어, 보기가 안 좋아 Stream exception 키워드로 검색하니 포비님이 조언해주신 글이 있더라구요. ㅎㅎ 그 글보고 한번 구현해 봤습니다. (public은.. private으로 해야 했는데 미처 보지 못했네요 ㅠ)

소현님의 리뷰를 보고 생각해보니 지금 상황에서 트랜잭션 내부에서 Exception을 상속한 객체를 사용 할 경우 Rollback에 문제가 생길 것이 다분하고, Exception 대신 RuntimeException을 상속받는다면, 이슈가 된 코드가 다 무의미 해 진다는 것을 깨닫고 수정 완료했습니다.

수정하면서 역량이 부족하다 느껴 소현님의 견해가 궁금한 부분 질문드립니다.

1. 트랜잭션 내부 Checked Exception이 발생할 경우, 많은 경우가 Checked Exception는 비지니스적 의미를 담고 있기에 스프링은 rollback 전략 대신 commit을 하고 내용을 진행 시킨다고 알고있습니다. 다만 실무에서 트랜잭션 내부 익셉션을 다룰 때, Unchecked Exception와 Checked Exception 중 무엇을 쓸지 고민하고 정하시는지, 혹시 Checked Exception을 사용했다면 어떤 경우에 사용했을지 궁금합니다.

ssosso reviewed (approved):
아하, 그러셨군요~ 의도하신 바를 이해했습니다:)
질문주신 내용은 effective java 서적에 매우 자세하게 관련 내용이 나와있습니다:) 제가 설명을 드려도 단편적이라 서적을 읽어보시는게 더 좋을거 같아요!
제 경험상 서적에 나와있듯이 여러 이유로 인해 실무에서는 checked exception을 잘 사용하지 않는 편입니다:)
